drbg/src/hashdrbg.rs

```diff
@@ -0,0 +1,220 @@
+use crate::{
+    helpers::{slice_to_u896, truncate_from_start},
+    props::{
+        MAX_BYTES_ADDITIONAL_INPUT, MAX_BYTES_PERSONALIZATION_STRING, MAX_BYTES_PER_REQUEST,
+        MAX_REQUESTS_BETWEEN_RESEEDS,
+    },
+    HashDrbgProps,
+};
+use core::iter::once;
+use crypto_bigint::{prelude::ArrayEncoding, U896};
+pub use digest;
+use digest::{generic_array::GenericArray, typenum::Unsigned, Digest, Output};
+use tpm2_rs_server::platform::crypto::{
+    drbg_helpers::{next_u32_via_fill, next_u64_via_fill},
+    Drbg, DrbgError,
+};
+
+/// [HashDrbg] is a generic software-based construct for creating
+/// a Hash DRBG compliant with
+/// [NIST SP 800-90Ar1](http://dx.doi.org/10.6028/NIST.SP.800-90Ar1).
+///
+/// It is based on the specifications outlined in _10.1 DRBG Mechanisms Based
+/// on Hash Functions_. This DRBG mechanism implementation uses the
+/// highest security strength of available in the hash function. Thus
+/// `requested_security_strength` parameter is omitted from arguments to
+/// `instantiate` function as well as from the internal state.
+///
+/// ## Notes
+///  `Hash` is expected to be cryptographically secure hash function.
+///
+pub struct HashDrbg<Hash: Digest + HashDrbgProps> {
+    /// _10.1.1.1 1. a._ A value (V) of _seedlen_ bits that is updated during
+    /// each call to the DRBG.
+    pub(crate) v: GenericArray<u8, Hash::SeedLenBytes>,
```

Comment:
The only reason I can see for these being `pub(crate)` is testing. And if that's the case then the tests are in the wrong physical location (e.g. you should not need a `pub(crate)` visibility here).

A property of Rust visibility is that submodules can see the private elements of theis parent modules. So what you probably want here is this:

*  drbg/src/hashdrbg/mod.rs
*  drbg/src/hashdrbg/unit_tests.rs
*  drbg/src/hashdrbg/nist_tests.rs

And then from `mod.rs` you can declare the tests as submodules:

```
#[cfg(test)]
mod unit_tests;
#[cfg(test)]
mod nist_tests;
```

Reply:
If there's another reason I don't see then please let me know, just wanted to comment on this before I forgot.